### ISSUE TITLE
Normal maps with steep angles don't light properly with Light2D.

### ISSUE THREAD

**Issue by jitspoe** [godotengine/godot#43309]
**Godot version:** 3.2.3

**OS/device including version:** Windows 8.1, Nvidia GTX 970, GLES2

**Issue description:** Steep angles of normal maps don't light properly (divide by 0?).  Spots end up entirely unlit.
![image](https://user-images.githubusercontent.com/7751182/98101630-7abf8e80-1e60-11eb-89e2-28cebfeaa38f.png)

Seems the NORMAL itself is incorrect, as I attempted to do my own custom lighting in a material, and doing a dot product with NORMAL yielded bad values.  I was able to fix the problem in a material doing the following:
`NORMAL = texture(NORMAL_TEXTURE, UV).rgb * 2.0 - vec3(1.0);`

I guess that's different from whatever Godot does internally?

**Steps to reproduce:**  Generate a normal map that has surfaces nearly parallel to the camera direction and apply it to a sprite with lighting.  Example normal map:
![loop_render_normal](https://user-images.githubusercontent.com/7751182/98100577-0f28f180-1e5f-11eb-9804-52b3a2cb8211.png)

**Minimal reproduction project:**
[test_bad_2d_normal_.zip](https://github.com/godotengine/godot/files/5487208/test_bad_2d_normal_.zip)

**Comment by lawnjelly** [comment]
Yes I can see this happening too. As the normal map was quite dithered I smoothed it in the gimp, but still something appears to be saturating. I wonder if it has to do with the normal map compression (I think it stores it in 2 channels).

**Comment by jitspoe** [comment]
Yeah, there's just some roughness to it, but even if I make something completely smooth, it still has the issue (just large chunks of areas are unlit instead of noisy values).  My guess is that the normal map is ignoring the blue channel and trying to reconstruct it (as happens in the 3D renderer), but normal values that are near or beyond perpendicular to the surface cause it to divide by 0 or something extreme.  I noticed when I did a dot product with 0, 1, 0 in a uniform variable, I was getting the same issue, but if I just did NORMAL.y, I did not, which threw me for a while until I figured out that the Z component was a massive negative value, and, for whatever reason, the vec3(0.0, 1.0, 0.0) wasn't exactly 0 for Z (??????!!!!), but only if it was a uniform variable.  Not if it was declared within the function.  If I declared a variable in the function with a value of like 0.00001 for z, it would also have the issue.  I think somehow, some of the components are ending up with like negative infinite values.
